code/modules/mob/living/living_say.dm

```diff
@@ -332,10 +335,19 @@ GLOBAL_LIST_INIT(message_modes_stat_limits, list(
 	var/whisper_range = 0
 	var/is_speaker_whispering = FALSE
 	if(message_mods[WHISPER_MODE]) //If we're whispering
-		whisper_range = EAVESDROP_EXTRA_RANGE
+		whisper_range = 6 // Needed for good hearing trait. The actual filtering for whispers happens at the /mob/living/Hear proc
+		// 6 is the value needed to allow people with GOOD_HEARING to hear at the default message range of 7. This range doesn't
+		// actually have a define, nor does the range for whispers
 		is_speaker_whispering = TRUE
 
-	var/list/listening = get_hearers_in_view(message_range + whisper_range, source)
+	var/list/in_view = dview(message_range + whisper_range, get_turf(source))
+	var/list/listening = get_hearers(message_range + whisper_range, source)
+
+	// Pre-process listeners to account for line-of-sight
+	for(var/atom/movable/listening_movable as anything in listening)
+		if(!(listening_movable in in_view) && !HAS_TRAIT(listening_movable, TRAIT_XRAY_HEARING))
+			listening.Remove(listening_movable)
```

Comment:
I'm a little worried about how performant this is compared to `get_hearers_in_view` which seems to be a proc someone has taken a lot of effort to optimise, that said I also don't know how you would marry the two 😔

Reply:
doing a range 13 dview with no for each filtering is terrible (doing for(var/atom/movable/thing in view(range)) is way faster than var/list/atoms = view(range) because the for loop only looks for /atom/movable but the list includes every atom), the change is mostly fine if they change that to get_hearers_in_view(). 

send_speech() is expensive per call but it doesnt get called very much, and the new proc is cheaper than get_hearers_in_view() with the changes i suggested. the main thing that makes me uncomfortable is adding 5 more range to the radius of get_hearers_in_view() (8^2 = 64 turfs for view() to look through at most, 13^2 = 169), but large open spaces are rare so i think its fine

Reply:
To clarify for other people, Kylerace and I discussed this over discord. This does not actually do anything with range 13, just range 7, and this fact has been made more clear through the use of new defines.